allow extensions to use `!` and `^` as special characters

QuietMisdreavus opened the pull request:
While implementing updates for https://github.com/apple/swift-cmark/pull/28, and while verifying changes for https://github.com/apple/swift-cmark/pull/29, i noticed that some special characters weren't being handled properly, and couldn't be used by extensions. This updates the `parse_inline` function to allow `^` and `!` to be used as special characters by extensions. This also allows footnotes to work with our implementation of inline attributes, allowing the full test suite to pass.

QuietMisdreavus committed:
allow extensions to use `!` and `^` as special characters

QuietMisdreavus committed:
update Makefile with new include dir

QuietMisdreavus commented:
I split off a change from https://github.com/apple/swift-cmark/pull/28 into this branch so that `make test` could pass with the swift-cmark-gfm changes.

QuietMisdreavus commented:
I've rebased this PR atop the new updates in #29.

daniel-grumberg commented on src/inlines.c:
```diff
@@ -1398,38 +1558,55 @@ static int parse_inline(cmark_parser *parser, subject *subj, cmark_node *parent,
+      push_bracket(subj, IMAGE, new_inl);
     }
     break;
-  default:
```
What happenned to the default case? Did you just push it outside the switch?

QuietMisdreavus replied:
Yes. I had to move the `try_extensions` call outside of this switch, so that it could be tried with characters that are part of the built-in grammar (specifically `^` and `[`). That's also why i modified the `!` and `^` branches to peek for their syntax instead of eagerly advancing the pointer.

daniel-grumberg commented on src/inlines.c:
```diff
@@ -1563,43 +1564,49 @@ static int parse_inline(cmark_parser *parser, subject *subj, cmark_node *parent,
   case '!':
-    advance(subj);
-    if (peek_char(subj) == '[' && peek_char_n(subj, 1) != '^') {
+    if (peek_char_n(subj, 1) == '[' && peek_char_n(subj, 2) != '^') {
```
Maybe add a comment saying that this case is not effectively "![(!^)"

QuietMisdreavus committed:
review comments

- generalize the special-character-inline-creating section
- add comments about looking for images and attributes specificall

daniel-grumberg reviewed (approved):
LGTM after offline discussion
